[8.x] Pass $key to closure in Collection and LazyCollection's reduce method as well

mokhosh opened the pull request:
Following the discussion we had here https://github.com/laravel/framework/pull/35839 I've removed `reduceWithKeys` and changed the `reduce` method to pass `$key` to its closure as well as  `$value`.

The previous test is there. I've just added another test for closures that accept the key, since I thought you might want to have one assertion per test. If you prefer to have one test per method and you don't care how many assertions are in one test, we can merge the two tests.

mokhosh committed:
add reduce with keys to collections

mokhosh committed:
add reduce with keys to lazy collections

mokhosh committed:
add test for reduce with keys

mokhosh committed:
fix style

mokhosh committed:
merge reduceWithKeys into the existing reduce method

mokhosh committed:
Merge branch '8.x' into 8.x

JosephSilber commented on tests/Support/SupportCollectionTest.php:
```diff
@@ -3601,7 +3601,7 @@ public function testReduceWithKeys($collection)
             'baz' => 'qux',
         ]);
-        $this->assertEquals('foobarbazqux', $data->reduceWithKeys(function ($carry, $element, $key) {
+        $this->assertEquals('foobarbazqux', $data->reduce(function ($carry, $element, $key) {
             return $carry .= $key.$element;
         }));
```
1. Since we're still keeping `reduceWithKeys` around till 9.x (because it was already tagged), we shouldn't remove its test.
2. You can just add the `$key` to the other test. No need for 2 separate tests.
3. As discussed on that other PR, you move the `reduce` method into the `EnumeratesValues` trait.

mokhosh replied:
Will do, thanks for the comment

mokhosh committed:
remove reduce and reduceWithKeys from Collection

mokhosh committed:
remove reduce and reduceWithKeys from LazyCollection

mokhosh committed:
add reduce and reduceWithKeys to EnumeratesValues

mokhosh committed:
add test for reduce with keys and reduceWithKeys

mokhosh committed:
Merge branch '8.x' into 8.x

mokhosh commented:
I think we're there.

1. kept the test for 8.x and removed it in the pull request I made for 9.x
2. added test with key to the same test
3. moved both methods to `EnumeratesValues`

mokhosh commented:
Doesn't StyleCI remove those 4 spaces automatically?

driesvints commented:
@mokhosh upon merging, not for forks.

mokhosh commented:
We're good then? Or do i fix style manually?

driesvints commented:
@mokhosh it's always better if everything's green of course but StyleCI issues will get fixed automatically upon merging.
